pym/bob/cmds/build/builder.py

```diff
@@ -126,6 +129,79 @@ def stop(self):
         os.write(self.__w_make, bytes(0x30))
         self.thread.join()
 
+
+class JobServerSemaphore:
+    def __init__(self, value, fds):
+        if sys.platform == "win32":
+            self.__sem = asyncio.BoundedSemaphore(value)
+        else:
+            self.__sem = asyncio.Semaphore(0)
+            self.__lock = threading.Lock()
+            self.__waitersCnt = 0
+            self.__jobsAquired = 0
+            self.__fds = fds
+            fcntl.fcntl(self.__fds[0], fcntl.F_SETFL,
+                    fcntl.fcntl(self.__fds[0], fcntl.F_GETFL) | os.O_NONBLOCK)
+
+    @staticmethod
+    def jobavailableCallback(self):
+        ret = None
+        try:
+            ret = os.read(self.__fds[0], 1)
+        except BlockingIOError:
+            pass
+        if ret:
+            with self.__lock:
+                self.__waitersCnt -= 1
+                if self.__waitersCnt == 0:
+                    asyncio.get_event_loop().remove_reader(self.__fds[0])
+            self.__sem.release()
+
+    async def acquire(self):
+        if sys.platform == "win32":
+            await self.__sem.acquire()
+        else:
+            ret = None
+            try:
+                ret = os.read(self.__fds[0], 1)
+            except BlockingIOError:
+                pass
+            if not ret:
+                with self.__lock:
+                    if self.__waitersCnt == 0:
+                        asyncio.get_event_loop().add_reader(self.__fds[0],
+                            JobServerSemaphore.jobavailableCallback, self)
+                    self.__waitersCnt += 1
+                await self.__sem.acquire()
+            with self.__lock:
+                self.__jobsAquired += 1
+
+    async def __aenter__(self):
+        await self.acquire()
+        return None
+
+    def release(self):
+        if sys.platform == "win32":
+            self.__sem.release()
+        else:
+            with self.__lock:
+                if self.__jobsAquired != 0:
+                    os.write(self.__fds[1], bytes(1))
+                    self.__jobsAquired -= 1
+
+    async def __aexit__(self, exc_type, exc, tb):
+        if sys.platform == "win32":
+            await self.__sem.__aexit__(exc_type, exc, tb)
+        else:
+            with self.__lock:
+                if self.__waitersCnt != 0:
+                    self.__waitersCnt = -1;
```

Comment:
:facepalm: :facepalm: :facepalm: :facepalm: 
it took me several hours to find the reason why bob sometimes didn't finish... 
:angry: :angry: :angry: :angry: